### PR TITLE
Refactor immutability to trait

### PR DESCRIPTION
**Pull request by Tito-Kati**
## 📚 Description

I turn immutability into a trait with the goal of adding more traits in the future (spoilers).

## 🔖 Changes

- Now immutability is a trait
- Slightly modified the immutability exception message to include the name of the immutable class

### CODE REVIEW

File: tests/Unit/Traits/ImmutableTest.php [hunk 1]
```diff
@@ -2,25 +2,27 @@
...
 use PHPUnit\Framework\TestCase;
 
-final class ImmutableArrayOfTest extends TestCase
+final class ImmutableTest extends TestCase
 {
     public function testImmutabilityOfSet(): void
     {
         $test = new ValidScalarImmutableArrayOf(['test']);
         $this->expectException(ImmutabilityException::class);
+        $this->expectExceptionMessage('ArrayOfTest\Unit\Fixtures\ValidScalarImmutableArrayOf objects are immutable.');
```
[the remark this answers is not in the document]
Comment (Tito-Kati) [reply tests/Unit/Traits/ImmutableTest.php:17]: I was trying to make sure that the class name also appears, would it be better to do it another way?

File: src/Exceptions/ImmutabilityException.php [hunk 2]
```diff
@@ -8,8 +8,8 @@
 
 final class ImmutabilityException extends Exception
 {
-    public function __construct()
+    public function __construct(string $className)
     {
-        parent::__construct('ArrayOf objects are immutable.');
+        parent::__construct("${className} objects are immutable.");
```
[the remark this answers is not in the document]
Comment (Tito-Kati) [reply src/Exceptions/ImmutabilityException.php:13]: https://www.php.net/manual/en/language.types.string.php#language.types.string.parsing doc?

[the remark this answers is not in the document]
Comment (Tito-Kati) [reply src/Exceptions/ImmutabilityException.php:13]: https://www.php.net/manual/en/language.types.string.php#language.types.string.parsing.complex ok xD

File: tests/Unit/Traits/ImmutableTest.php (hunk 1 above)
[the remark this answers is not in the document]
Comment (Chemaclass) [reply tests/Unit/Traits/ImmutableTest.php:17]: What do you think about `ValidScalarImmutableArrayOf objects are immutable.`?
This way the test won't be dependent on the namespace, but on the class name and the error message.

**Review by JesusValera — APPROVED** [review]

**Review by Chemaclass — APPROVED** [review]

**Pull request merged**